### PR TITLE
refactor screenshare to allow audio

### PR DESCRIPTION
**Pull request by czerwiukk**
## Description

Implements use screenshare hook.

## Motivation and Context

Why is this change required? What problem does it solve? If it fixes an open
issue, please link to the issue here.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to
      not work as expected)

### CODE REVIEW

File: packages/react-client/src/screenShareTrackManager.ts [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+  const [stream, setStream] = useState<MediaStream | null>(null);
+  const [trackIds, setTrackIds] = useState<{ videoId: string; audioId?: string } | null>(null);
+
+  useEffect(() => {
+    if (!stream) return;
+    const { video, audio } = getTracks(stream);
+
+    const getTrackEndedHandler = (track: MediaStreamTrack) => () => {
+      stream.getTrackById(track.id)?.stop();
+    };
```
Comment (kamil-stasiak) [inline packages/react-client/src/screenShareTrackManager.ts:22]: Will this still work if we simplify it?

```suggestion
    const trackEndedHandler = (track: MediaStreamTrack) => {
      stream.getTrackById(track.id)?.stop();
    };
```

Or do we require a separate handler for audio and video?

File: packages/react-client/src/utils/tracks.ts [hunk 2]
```diff
@@ -0,0 +1,41 @@
+import { FishjamClient, TrackContext } from "@fishjam-dev/ts-client";
+import { Track } from "../state.types";
+
+export const getRemoteTrackContext = <PeerMetadata, TrackMetadata>(
```
Comment (kamil-stasiak) [inline packages/react-client/src/utils/tracks.ts:4]: Could be private
```suggestion
const getRemoteTrackContext = <PeerMetadata, TrackMetadata>(
```

File: packages/react-client/src/utils/tracks.ts [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+  trackId: track.trackId,
+  stream: track.stream,
+  simulcastConfig: track.simulcastConfig || null,
+  encoding: track.encoding || null,
+  vadStatus: track.vadStatus,
+  track: track.track,
+  metadataParsingError: track.metadataParsingError,
+});
+
+export const getRemoteTrack = <PeerMetadata, TrackMetadata>(
```
Comment (kamil-stasiak) [inline packages/react-client/src/utils/tracks.ts:34]: I would move this method into the `FishjamClient` along with `getRemoteTrack` and `getRemoteTrackContext` (as private methods)

Comment (kamil-stasiak) [inline packages/react-client/src/utils/tracks.ts:34]: `getRemoteTrack` from `Client.ts` is never called, so you can delete it. 

If `getRemoteTrack` description is still valid, consider moving it to `getRemoteTrackContext`

**Review by kamil-stasiak — APPROVED** [review]

File: packages/react-client/src/screenShareTrackManager.ts (hunk 1 above)
Comment (czerwiukk) [reply packages/react-client/src/screenShareTrackManager.ts]: no :)

File: examples/react-client/use-camera-and-microphone/src/MainControls.tsx [hunk 4]
```diff
@@ -427,17 +418,25 @@ export const MainControls = () => {
...
-                  stream={audio?.stream}
-                  trackId={audio?.track?.id ?? null}
+                  stream={audio.stream}
+                  trackId={audio.track.id ?? null}
                 />
               )}
-              {screenShare?.track?.kind === "video" && (
+
+              {screenShare.videoTrack && (
                 <VideoPlayer stream={screenShare?.stream} />
```
Comment (kamil-stasiak) [inline examples/react-client/use-camera-and-microphone/src/MainControls.tsx:431]:
```suggestion
                <VideoPlayer stream={screenShare.stream} />
```

File: examples/react-client/use-camera-and-microphone/src/MainControls.tsx [hunk 5]
```diff
@@ -427,17 +418,25 @@ export const MainControls = () => {
...
               )}
+
+              {screenShare.audioTrack && (
+                <AudioVisualizer
+                  trackId={screenShare.audioTrack.id}
+                  stream={screenShare?.stream}
```
Comment (kamil-stasiak) [inline examples/react-client/use-camera-and-microphone/src/MainControls.tsx:437]:
```suggestion
                  stream={screenShare.stream}
```

File: packages/react-client/src/trackManager.ts [hunk 6]
```diff
@@ -69,11 +46,11 @@ export class TrackManager<PeerMetadata, TrackMetadata> implements GenericTrackMa
 
     if (!media || !media.stream || !media.track) throw Error("Device is unavailable");
 
-    const track = this.getRemoteTrack(media.track.id);
+    const track = getRemoteOrLocalTrack(this.tsClient, this.currentTrackId);
 
     if (track) return track.trackId;
 
-    // see `getRemoteTrack()` explanation
+    // see `getRemoteOrLocalTrack()` explanation
```
Comment (kamil-stasiak) [inline packages/react-client/src/trackManager.ts:53]:
```suggestion
    // see `getRemoteOrLocalTrackContext()` explanation
```

File: packages/react-client/src/screenShareTrackManager.ts [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+  const audio = stream.getAudioTracks()[0] ?? null;
+
+  return { video, audio };
+};
+
+export const useScreenShare = <PeerMetadata, TrackMetadata>(
+  // these arguments will be removed after we get rid of create() fn, we will just use the context
+  [state, setState]: [ScreenshareState, React.Dispatch<React.SetStateAction<ScreenshareState>>],
+  tsClient: FishjamClient<PeerMetadata, TrackMetadata>,
+) => {
```
Comment (kamil-stasiak) [inline packages/react-client/src/screenShareTrackManager.ts:17]: WDYT about using an explicit return type?

**Review by kamil-stasiak — APPROVED** [review]

**Pull request merged**